flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -275,7 +275,7 @@ private void handleStartDispatcherServicesException(Exception e) throws Exceptio
     @Override
     public CompletableFuture<Void> onStop() {
         log.info("Stopping dispatcher {}.", getAddress());
-
+        super.onStop();
```

Comment:
Honestly, I don't like this solution since it looks pretty fragile. I tried to move closing mainThreadExecutor to `RpcEndpoint#internalCallOnStop` which I like more but unfortunately, it is impossible to do so because we have one more `Executor` in `FencedRpcEndpoint`.
In my opinion, we should rethink the management of resources here. Perhaps we should register all resources in one `CloseableRegistry` and then close it in one place(for example, in `RpcEndpoint#internalCallOnStop` after `onStop` future)

Reply:
Thanks @akalash. I like the idea of `ClosableRegistry` to close all resources in one place. In fact it confuses me a little while there're `stop`, `onStop`, `internalCallOnStop` and `closeAsync` in `RpcEndpoint`. I think it's very good and necessary to close all the resources in one place

Reply:
Actually, I don't have a strong preference here. In the current implementation, we need to claim in the JavaDoc of `RpcEndpoint#onStop` that the `super.onStop()` should always be called.
I agree it looks fragile and `AutoCloseableRegistry` sounds like a good idea. However, one can also complain that people might forget to register the resources to the `AutoCloseableRegistry`.
Either way, I think it is important to document it in a prominent position.

Reply:
I thought about it a little more. Right now, I think that in general, we need to have one pattern for handling resources but it is not clear what it should be for now and I agree that `AutoCloseableRegistry` also has a problem. So maybe it makes sense to leave the changes as is at least for now.
If nobody minds of these changes let's wait for the test result and then merge it. @KarmaGYZ

Reply:
Sounds good to me. Would you like to file a ticket for the rethink of resource cleanup protocol in `RpcEndpoint`?

Reply:
Perhaps, it makes sense. I need to think about how to formulate the problem(Highly likely, it is not only about `RpcEndpoint` but for the similar classes).